packages/runtime/container-runtime/src/channelCollection.ts

```diff
@@ -865,47 +861,105 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
+			const envelope = contents as IEnvelope;
+			const address = envelope.address;
+			const context = this.contexts.get(address);
+
+			// If the data store has been deleted, log an error and ignore this message. This helps prevent document
+			// corruption in case a deleted data store accidentally submitted an op.
+			if (this.checkAndLogIfDeleted(address, context, "Changed", "processFluidDataStoreOp")) {
+				continue;
+			}
 
-		if (context === undefined) {
-			// Former assert 0x162
-			throw DataProcessingError.create(
-				"No context for op",
-				"processFluidDataStoreOp",
-				message,
-				{
-					local,
-					messageDetails: JSON.stringify({
-						type: message.type,
-						contentType: typeof message.contents,
-					}),
-					...tagCodeArtifacts({ address }),
-				},
+			if (context === undefined) {
+				// Former assert 0x162
+				throw DataProcessingError.create(
+					"No context for op",
+					"processFluidDataStoreOp",
+					message as ISequencedDocumentMessage,
+					{
+						local,
+						messageDetails: JSON.stringify({
+							type: message.type,
+							contentType: typeof contents,
+						}),
+						...tagCodeArtifacts({ address }),
+					},
+				);
+			}
+
+			const { type: contextType, content: contextContents } =
+				envelope.contents as FluidDataStoreMessage;
+			// If the address or type of the message changes while processing the message, send the current bunch.
+			if (
+				currentMessageState !== undefined &&
+				(currentMessageState.address !== address || currentMessageState.type !== contextType)
+			) {
+				sendBunchedMessages();
+			}
+			currentMessageContents.push({
+				contents: contextContents,
+				...restOfMessageContents,
+			});
+			currentMessageState = { address, type: contextType };
+
+			// Notify that a GC node for the data store changed. This is used to detect if a deleted data store is
+			// being used.
+			this.gcNodeUpdated({
+				node: { type: "DataStore", path: `/${address}` },
+				reason: "Changed",
+				timestampMs: message.timestamp,
+				packagePath: context.isLoaded ? context.packagePath : undefined,
+			});
+
+			detectOutboundReferences(
+				envelope.address,
+				contextContents,
+				(fromPath: string, toPath: string) =>
+					this.parentContext.addedGCOutboundRoute(fromPath, toPath, message.timestamp),
 			);
 		}
 
-		context.process(message, local, localMessageMetadata);
-
-		// Notify that a GC node for the data store changed. This is used to detect if a deleted data store is
-		// being used.
-		this.gcNodeUpdated({
-			node: { type: "DataStore", path: `/${address}` },
-			reason: "Changed",
-			timestampMs: message.timestamp,
-			packagePath: context.isLoaded ? context.packagePath : undefined,
-		});
+		// Process the last bunch of messages, if any. Note that there may not be any messages in case all of them are
+		// ignored because the data store is deleted.
+		if (currentMessageState !== undefined) {
+			sendBunchedMessages();
```

Comment:
Is it easier to make sendBunchedMessages() deal with empty state and not check in both place before calling it?

Reply:
I moved it inside `sendBunchedMessages`. Not sure if it's any better but either way is fine :-)